repositories/decision_phantoms_repository.go

```diff
@@ -0,0 +1,121 @@
+package repositories
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/checkmarble/marble-backend/models"
+	"github.com/checkmarble/marble-backend/models/ast"
+	"github.com/checkmarble/marble-backend/repositories/dbmodels"
+	"github.com/checkmarble/marble-backend/utils"
+	"github.com/cockroachdb/errors"
+	"github.com/google/uuid"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+)
+
+type DecisionPhantomUsecaseRepository interface {
+	StorePhantomDecision(
+		ctx context.Context,
+		exec Executor,
+		decision models.PhantomDecision,
+		organizationId string,
+		testRunId string,
+		newPhantomDecisionId string) error
+
+	GetTestRunIterationByScenarioId(ctx context.Context, exec Executor, scenarioID string) (models.ScenarioIteration, error)
+}
+
+func (repo *MarbleDbRepository) StorePhantomDecision(
+	ctx context.Context,
+	exec Executor,
+	decision models.PhantomDecision,
+	organizationId string,
+	testRunId string,
+	newPhantomDecisionId string,
+) error {
+	tracer := utils.OpenTelemetryTracerFromContext(ctx)
+	ctx, span := tracer.Start(
+		ctx,
+		"DecisionPhantomRepository.StorePhantomDecision.store_phantom_decision",
+		trace.WithAttributes(attribute.String("phantom_decision_id", newPhantomDecisionId)))
+	defer span.End()
+	if err := validateMarbleDbExecutor(exec); err != nil {
+		return err
+	}
+	err := ExecBuilder(
+		ctx,
+		exec,
+		NewQueryBuilder().Insert(dbmodels.TABLE_PHANTOM_DECISIONS).
+			Columns(
+				"id",
+				"org_id",
+				"created_at",
+				"outcome",
+				"scenario_id",
+				"scenario_iteration_id",
+				"score",
+				"test_run_id",
+			).
+			Values(
+				newPhantomDecisionId,
+				organizationId,
+				decision.CreatedAt,
+				decision.Outcome.String(),
+				decision.ScenarioId,
+				decision.ScenarioIterationId,
+				decision.Score,
+				testRunId,
+			),
+	)
+	if err != nil {
+		return err
+	}
+
+	ctx, span = tracer.Start(
+		ctx,
+		"DecisionPhantomRepository.StorePhantomDecision.store_phantom__decision_rules",
+		trace.WithAttributes(attribute.String("phantom_decision_id", newPhantomDecisionId)))
+	defer span.End()
+	builderForRules := NewQueryBuilder().
+		Insert(dbmodels.TABLE_DECISION_RULES).
+		Columns(
+			"id",
+			"org_id",
+			"decision_id",
+			"score_modifier",
+			"result",
+			"error_code",
+			"rule_id",
+			"rule_evaluation",
+			"outcome",
+		)
+	for _, ruleExecution := range decision.RuleExecutions {
+		serializedRuleEvaluation, err := dbmodels.SerializeNodeEvaluationDto(ruleExecution.Evaluation)
+		if err != nil {
+			return errors.Wrap(err, fmt.Sprintf("rule(%s):", ruleExecution.Rule.Id))
+		}
+
+		builderForRules = builderForRules.
+			Values(
+				uuid.Must(uuid.NewV7()).String(),
+				organizationId,
+				newPhantomDecisionId,
+				ruleExecution.ResultScoreModifier,
+				ruleExecution.Result,
+				ast.AdaptExecutionError(ruleExecution.Error),
+				ruleExecution.Rule.Id,
+				serializedRuleEvaluation,
```

Comment:
related to comment above: this one could be removed.
Also, I'm wondering if we should have a separate decision_rules table for phantom decisions, since we have "split" the decisions table.
(Maybe not, since it doesn't have explicit FKs anyway)

Reply:
works for me!